### PR TITLE
ISPN-16231 Skip return value internally if Hot Rod request does not force return value.

### PR DESCRIPTION
**Pull request by jabolina**
https://issues.redhat.com/browse/ISPN-16231

Check on the CacheRequestProcessor so we can avoid creating a copy for every operation.

### CODE REVIEW

File: server/hotrod/src/main/java/org/infinispan/server/hotrod/CacheRequestProcessor.java [hunk 1]
```diff
@@ -234,8 +234,14 @@ void put(HotRodHeader header, Subject subject, byte[] key, byte[] value, Metadat
 
    private void putInternal(HotRodHeader header, AdvancedCache<byte[], byte[]> cache, byte[] key, byte[] value,
                             Metadata metadata, InfinispanSpan<CacheEntry<?, ?>> span) {
-      cache.putAsyncEntry(key, value, metadata)
-            .whenComplete((ce, throwable) -> handlePut(header, ce, throwable))
+      CompletionStage<CacheEntry<byte[], byte[]>> cs;
+      if (header.hasFlag(ProtocolFlag.ForceReturnPreviousValue)) {
+         cs = cache.putAsyncEntry(key, value, metadata);
+      } else {
+         cs = cache.putAsync(key, value, metadata)
```
Comment (wburns) [inline server/hotrod/src/main/java/org/infinispan/server/hotrod/CacheRequestProcessor.java:241]: This won't do much of anything. You want to invoke `cache.withFlags(Flag.IGNORE_RETURN_VALUES)`.

File: server/hotrod/src/main/java/org/infinispan/server/hotrod/CacheRequestProcessor.java [hunk 2]
```diff
@@ -234,8 +234,15 @@ void put(HotRodHeader header, Subject subject, byte[] key, byte[] value, Metadat
...
                             Metadata metadata, InfinispanSpan<CacheEntry<?, ?>> span) {
-      cache.putAsyncEntry(key, value, metadata)
-            .whenComplete((ce, throwable) -> handlePut(header, ce, throwable))
+      CompletionStage<CacheEntry<byte[], byte[]>> cs;
+      if (header.hasFlag(ProtocolFlag.ForceReturnPreviousValue)) {
+         cs = cache.putAsyncEntry(key, value, metadata);
+      } else {
+         cs = cache.withFlags(Flag.IGNORE_RETURN_VALUES)
+               .putAsync(key, value, metadata)
+               .thenApply(CompletableFutures.toNullFunction());
```
Comment (wburns) [inline server/hotrod/src/main/java/org/infinispan/server/hotrod/CacheRequestProcessor.java:243]: This should always be null as well

Comment (jabolina) [reply server/hotrod/src/main/java/org/infinispan/server/hotrod/CacheRequestProcessor.java:243]: The `putAsyncEntry` causes an additional synchronized entry clone/creation on the CallInterceptor.

https://github.com/infinispan/infinispan/blob/e8e0d45c869b2f8e22683e9ad09193163383b1c6/core/src/main/java/org/infinispan/interceptors/impl/CallInterceptor.java#L1156-L1161

I saw a few samples on the flame graph and decided to skip the method if it was unneeded.

Comment (wburns) [reply server/hotrod/src/main/java/org/infinispan/server/hotrod/CacheRequestProcessor.java:243]: Okay, I would suggest testing the two changes separately at least. Flag.IGNORE_RETURN_VALUES should be a much bigger deal as it avoids sending back the entire value payload for intra cluster communication on the network and an additional unmarshalling.

**Comment by infinispanrelease** [comment]
Image pushed for Jenkins build [#3](https://ci.infinispan.org/job/Infinispan/job/PR-12511/3/):
```
quay.io/infinispan-test/server:PR-12511
```

**Comment by infinispanrelease** [comment]
Image pushed for Jenkins build [#4](https://ci.infinispan.org/job/Infinispan/job/PR-12511/4/):
```
quay.io/infinispan-test/server:PR-12511
```

**Review by wburns — APPROVED** [review]

**Comment by wburns** [comment]
Integrated into main, thanks @jabolina !

**Pull request merged**